### PR TITLE
Diagnostig plots

### PR DESCRIPTION
**Pull request by JarnoHerr**
Made a start with the tool for diagnostic plots. Still needs to be adjusted for the input observations.

### CODE REVIEW

**Comment by MuellerSeb** [comment]
Some notes:
- put a `diagnostic_plot` method next to `plot` in the `PumpingTest` class:
  https://github.com/GeoStat-Framework/welltestpy/blob/5bdf033b7dbf3439862424ff23ba3f7b234d150d/welltestpy/data/testslib.py#L442
- this method should call a plotting routine placed in `plotter` called `diagnostic_plot_pump_test` similar to this method
  https://github.com/GeoStat-Framework/welltestpy/blob/5bdf033b7dbf3439862424ff23ba3f7b234d150d/welltestpy/tools/plotter.py#L266
- place the needed derivative routines in `processlib` (similar to `cooper_jacob_correction`):
  https://github.com/GeoStat-Framework/welltestpy/blob/5bdf033b7dbf3439862424ff23ba3f7b234d150d/welltestpy/process/processlib.py#L202
  - this routine should take a single observation class and return an array with the derivative (no inplace operations on the observations)
  - add a switch to select the method ("bourdet" for now)
- finally: always apply `black -l 79 .` (see [here](https://github.com/psf/black)) to format the code

File: welltestpy/process/processlib.py [hunk 1]
```diff
@@ -40,9 +40,7 @@ def normpumptest(pumptest, pumpingrate=-1.0, factor=1.0):
     pumptest.pumpingrate = pumpingrate
 
     for obs in pumptest.observations:
-        pumptest.observations[obs].observation *= (
-            factor * pumptest.rate / oldprate
-        )
+        pumptest.observations[obs].observation *= factor * pumptest.rate / oldprate
```
Comment (JarnoHerr) [inline welltestpy/process/processlib.py:43]: Result of the black reformatting done. I dont know if it worked correctly :)

Comment (MuellerSeb) [reply welltestpy/process/processlib.py:43]: I guess you didn't use `black -l 79 .` where `-l 79` bounds the line-length to 79 characters.

File: welltestpy/tools/plotter.py [hunk 2]
```diff
@@ -895,3 +895,44 @@ def plotsensitivity(
...
+    # setting variables
+    x = time
+    y = head
+    sx = time
+    sy = head
+    dx = time
+    dy = derivative
+
+    # plotting
+    fig, ax = plt.subplots(dpi=75, figsize=[7.5, 5.5])
```
Comment (MuellerSeb) [inline welltestpy/tools/plotter.py:926]: If you take a look at the other plotting routines, you see, that they share a certain signature, including `fig` and `ax` (when plot includes a single axis) to be passed to the routine. Please follow this. You could also try to imitate the `style="WTP"` to select the plotting style. Have a look at this for example:
https://github.com/GeoStat-Framework/welltestpy/blob/5bdf033b7dbf3439862424ff23ba3f7b234d150d/welltestpy/tools/plotter.py#L862

File: welltestpy/tools/diagnostic_plots.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import numpy as np
+
+from mpl_toolkits.mplot3d import Axes3D
+import matplotlib.pyplot as plt
+import matplotlib.ticker as ticker
+
+
+
+
+def smoothing_derivative(observation):
```
Comment (MuellerSeb) [inline welltestpy/tools/diagnostic_plots.py:26]: I guess this routine could be in `process`. But I like the idea to have a `tools/diagnostic_plots` submodule.

File: welltestpy/data/testslib.py [hunk 4]
```diff
@@ -467,6 +467,27 @@ def plot(self, wells, exclude=None, fig=None, ax=None, **kwargs):
             **kwargs
         )
 
+    def diagnostic_plot(self, observation, **kwargs):
```
Comment (MuellerSeb) [inline welltestpy/data/testslib.py:470]: observation should be the name of the observation well. You can get the observation itself afterwards by:
```python
def diagnostic_plot(self, well, **kwargs):
    observation = self.observations[well]
    ...
```

**Comment by MuellerSeb** [comment]
Maybe we can also add a `diagnostic_plot` method to the `Campaign` class taking the name of the pumping well and the observation well:
```python
def diagnostic_plot(self, pumping_test, observation_well, **kwargs):
    # check if this really is a pumping test
    # check if the well is present
    return self.tests[pumping_test].diagnostic_plot(well=observation_well, **kwargs)
```

File: welltestpy/process/processlib.py [hunk 5]
```diff
@@ -214,7 +208,7 @@ def filterdrawdown(observation, tout=None, dxscale=2):
         Scale of time-steps used for smoothing.
         Default: ``2``
     """
-    head, time = observation()
+    head, time = observation
```
Comment (JarnoHerr) [inline welltestpy/process/processlib.py:211]: This is a mistake

File: welltestpy/data/testslib.py [hunk 6]
```diff
@@ -467,6 +467,26 @@ def plot(self, wells, exclude=None, fig=None, ax=None, **kwargs):
...
         )
 
+    def diagnostic_plot(self, well, **kwargs):
+        """Make a diagnostic plot.
+
+
+        Parameters
+        ----------
+       observation : :class:`Variable`
+            Observation.
```
Comment (JarnoHerr) [inline welltestpy/data/testslib.py:477]: Docstring needs to be updated

File: welltestpy/process/processlib.py [hunk 7]
```diff
@@ -240,3 +246,44 @@ def filterdrawdown(observation, tout=None, dxscale=2):
...
+                                 / (np.log(time[i]) - np.log(time[i - 1]))
+                                 * (np.log(time[i + 1]) - np.log(time[i]))
+                         )
+                         + (
+                                 (head[i + 1] - head[i])
+                                 / (np.log(time[i + 1]) - np.log(time[i]))
+                                 * (np.log(time[i]) - np.log(time[i - 1]))
+                         )
+                 ) / ((np.log(time[i]) - np.log(time[i - 1])) + (np.log(time[i + 1]) - np.log(time[i])))
+            derhead[i] = dh
```
Comment (MuellerSeb) [inline welltestpy/process/processlib.py:288]: These indentations look strange. Is it formatted with `black -l 79 .`?

**Review by MuellerSeb — CHANGES_REQUESTED** [review]
This looks very good so for. Only some minor changes. Will test it myself later this week.

File: welltestpy/data/testslib.py [hunk 8]
```diff
@@ -467,6 +467,26 @@ def plot(self, wells, exclude=None, fig=None, ax=None, **kwargs):
             **kwargs
         )
 
+    def diagnostic_plot(self, wells, **kwargs):
+        """Make a diagnostic plot.
+
+
+        Parameters
+        ----------
+       wells : :class:`dict`
```
Comment (MuellerSeb) [inline welltestpy/data/testslib.py:476]: Please rename `wells` to `observation_well` to be in line with the routine in campaign.

File: welltestpy/data/testslib.py [hunk 9]
```diff
@@ -467,6 +467,26 @@ def plot(self, wells, exclude=None, fig=None, ax=None, **kwargs):
...
+            Dictionary containing the well classes sorted by name.
+
+        Notes
+        -----
+        This will be used by the Campaign class.
+        """
+        observation = self.observations[wells]
```
Comment (MuellerSeb) [inline welltestpy/data/testslib.py:483]: Please check if the given observation_well is actually present in `self.observations`.

File: welltestpy/process/processlib.py [hunk 10]
```diff
@@ -240,3 +246,47 @@ def filterdrawdown(observation, tout=None, dxscale=2):
         hout = np.interp(tout, time, head)
 
     return observation(time=tout, observation=hout)
+
+
+def smoothing_derivative(observation, bourdet=True):
```
Comment (MuellerSeb) [inline welltestpy/process/processlib.py:251]: Please use a keyword `method="bourdet"` with default value "bourdet".
Then check for the method and raise an error if it is unknown.

File: welltestpy/tools/diagnostic_plots.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    with plt.style.context(style):
+        if keep_fs:
+            plt.rcParams.update({"font.size": font_size})
+        fig, ax = plotter._get_fig_ax(fig, ax)
+        ax.scatter(x, y, color="red", label="drawdown")
+        ax.plot(sx, sy, c="red")
+        ax.plot(dx, dy, c="black", linestyle='dashed')
+        ax.scatter(dx, dy, c="black", marker="+",  label="time derivative")
+        ax.set_xscale("symlog")
+        ax.set_yscale("symlog")
```
Comment (MuellerSeb) [inline welltestpy/tools/diagnostic_plots.py:64]: We need to determine the bounds for the symlog to have meaningful cut-off values.

File: welltestpy/tools/diagnostic_plots.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        ax.set_ylim([0, head[-1]])
+        ax.set_xlim(time[0], time[-1])
+        ax.set_xlabel("$t$ in [s]", fontsize=12)
+        ax.set_ylabel("$h$ and $dh/dx$ in [m]", fontsize=12)
+        ax.set_yticks(np.arange(0, head[-1], 0.02))
```
Comment (MuellerSeb) [inline welltestpy/tools/diagnostic_plots.py:69]: these ticks need to be only powers of 10. We need to determine the orders of magnitude for the range of head values.

**Review by MuellerSeb — CHANGES_REQUESTED** [review]
This is looking better and better. But still room for improvement :wink:

**Comment by MuellerSeb** [comment]
And we still need a test for (at least) `smoothing_derivative`.

File: welltestpy/tools/diagnostic_plots.py (hunk 11 above)
Comment (JarnoHerr) [reply welltestpy/tools/diagnostic_plots.py:64]: Do you mean with linthresh values for the symlogs?

Comment (MuellerSeb) [reply welltestpy/tools/diagnostic_plots.py:64]: Yes. Maybe we could use arguments with standard values: (`linthresh_time=1`, linthresh_head=`1e-5`)
```python
ax.set_xscale("symlog", linthresh=linthresh_time)
ax.set_yscale("symlog", linthresh=linthresh_head)
```
Then we could also use this as the lower bound for the axis (to omit the linear part of the (sym)log-log plot).
For the upper bound, we need to determine the order of magnitude from head and time.

But what happens if you leave the bound settings untouched? Matplotlib is quite smart with this.

File: welltestpy/tools/diagnostic_plots.py (hunk 12 above)
Comment (JarnoHerr) [reply welltestpy/tools/diagnostic_plots.py:69]: i tried to find a solution for this but the scales are to small to plot naturally and i cant find a way to not hard code the ticks.

Comment (JarnoHerr) [reply welltestpy/tools/diagnostic_plots.py]: I found a bit of a dirty solution but i couldnt find another way.

File: welltestpy/data/testslib.py [hunk 13]
```diff
@@ -467,25 +467,28 @@ def plot(self, wells, exclude=None, fig=None, ax=None, **kwargs):
...
 
-    def diagnostic_plot(self, wells, **kwargs):
+    def diagnostic_plot(self, observation_well, **kwargs):
         """Make a diagnostic plot.
 
 
         Parameters
         ----------
-       wells : :class:`dict`
+       observation_well : :class:`dict`
```
Comment (MuellerSeb) [inline welltestpy/data/testslib.py:476]:
```
:class:`str`
```

File: welltestpy/process/processlib.py [hunk 14]
```diff
@@ -248,16 +248,16 @@ def filterdrawdown(observation, tout=None, dxscale=2):
...
 
     Parameters
     ----------
     observation : :class:`welltestpy.data.Observation`
         The observation to calculate the derivative.
 
-    bourdet : :class:`bool`, optional
-        Optional switch to access the Bourdet time derivative
+    method : :class:`bool`, optional, default = bourdet
+        Optional switch to access the method to calculate the time derivative
```
Comment (MuellerSeb) [inline welltestpy/process/processlib.py:260]:
```
method : :class:`str`, optional
    Optional switch to access the method to calculate the time derivative.
    Default: "bourdet"
```

File: welltestpy/process/processlib.py [hunk 15]
```diff
@@ -270,7 +270,7 @@ def smoothing_derivative(observation, bourdet=True):
     time = np.array(time, dtype=float).reshape(-1)
     derhead = np.zeros(len(head))
     t = np.arange(len(time))
-    if bourdet is True:
+    if method is "bourdet":
```
Comment (MuellerSeb) [inline welltestpy/process/processlib.py:273]: `if method == "bourdet":`

File: welltestpy/process/processlib.py [hunk 16]
```diff
@@ -290,3 +290,7 @@ def smoothing_derivative(observation, bourdet=True):
             )
             derhead[i] = dh
         return derhead
+    else:
+        raise ValueError(
+            f"smoothing_derivative: method '{method}' could not be found!"
```
Comment (MuellerSeb) [inline welltestpy/process/processlib.py:295]: Rephrase: "could not be found" -> "unknown"

**Review by MuellerSeb — CHANGES_REQUESTED** [review]
Minor changes.

File: welltestpy/tools/diagnostic_plots.py [hunk 17]
```diff
@@ -58,15 +66,40 @@ def diagnostic_plot_pump_test(observation, fig=None, ax=None,plotname=None,style
...
+            1e5,
+        ]:
+            if i > min(head[0], derivative[0]):
+                if i < max(head[-1], derivative[-1]):
+                    yticks.append(i)
+                else:
+                    continue
+            else:
+                continue
+        ax.set_yticks(yticks)
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply welltestpy/tools/diagnostic_plots.py:100]: https://raw.githubusercontent.com/GeoStat-Examples/welltestpy-field-site-analysis/master/results/03_unconfined/horkheim_p05p44_analysis.png
Here this is only done for `head` (ignoring the range of the `derivative`).

[the remark this answers is not in the document]
Comment (MuellerSeb) [reply welltestpy/tools/diagnostic_plots.py:100]: And I guess we also need to use `np.abs(head)` to cut out negative values.

[the remark this answers is not in the document]
Comment (JarnoHerr) [reply welltestpy/tools/diagnostic_plots.py:100]: Returns "OverflowError: cannot convert float infinity to integer" min_e will try to get the log10 of 0.

[the remark this answers is not in the document]
Comment (MuellerSeb) [reply welltestpy/tools/diagnostic_plots.py:100]: Ok, then we need to cut this off:
```python
max_e = int(np.ceil(np.log10(max_v)))
if min_v < linthresh_head:
    ax.set_ylim(0, 10 ** max_e)
else:
    min_e = int(np.floor(np.log10(min_v)))
    ax.set_ylim(10 ** min_e, 10 ** max_e)
```

[the remark this answers is not in the document]
Comment (JarnoHerr) [reply welltestpy/tools/diagnostic_plots.py:100]: It works. But when the head values are small the plotted lines remain in the bottom part of the graph so that trends are not well shown.

File: welltestpy/data/testslib.py [hunk 18]
```diff
@@ -467,6 +467,29 @@ def plot(self, wells, exclude=None, fig=None, ax=None, **kwargs):
...
         )
 
+    def diagnostic_plot(self, observation_well, **kwargs):
+        """Make a diagnostic plot.
+
+
+        Parameters
+        ----------
+       observation_well : :class:`str`
+            Dictionary containing the well classes sorted by name.
```
Comment (MuellerSeb) [inline welltestpy/data/testslib.py:477]: Still wrong doc.

File: welltestpy/tools/diagnostic_plots.py [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+        Default: None.
+    style : str, optional
+        Plot style.
+        Default: "WTP".
+
+     Returns
+     ---------
+     Diagnostic plot
+    """
+    derivative = processlib.smoothing_derivative(observation,method)
```
Comment (MuellerSeb) [inline welltestpy/tools/diagnostic_plots.py:65]: apply `black -l 79 .` again.

**Review by MuellerSeb — CHANGES_REQUESTED** [review]
Looking better and better. Could you provide a resulting plot for the current implementations.
We would also need a small example in the `examples` folder using this on the generated campaign:
```python
import welltestpy as wtp

campaign = wtp.load_campaign("Cmp_UFZ-campaign.cmp")
campaign.diagnostic_plot("well_0", "well_1")
```

File: welltestpy/data/testslib.py [hunk 20]
```diff
@@ -176,6 +176,20 @@ def make_steady(self, time="latest"):
...
+        Parameters
+        ----------
+        wells : :class:`list`, optional
+            List of wells, to check the observation state at. Default: all
+        method : ...
+        """
+        wells = self.observationwells if wells is None else list(wells)
+
+        # loop over wells and apply cooper_jacob correction (check for method in if branches)
+
```
Comment (MuellerSeb) [inline welltestpy/data/testslib.py]: Why is that here?

File: welltestpy/tools/diagnostic_plots.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply welltestpy/tools/diagnostic_plots.py]: This is then controllable with `linthresh_head` or am I wrong?

[the remark this answers is not in the document]
Comment (JarnoHerr) [reply welltestpy/tools/diagnostic_plots.py]: I can't seem to change the plot when changing the linthresh_head. I think it has something to do with the np.ceil which creates, for the smallest values,  always a max of 10^0. This results in the too large y scale.

[the remark this answers is not in the document]
Comment (JarnoHerr) [reply welltestpy/tools/diagnostic_plots.py]: The 01_create has also a negative pumping rate which messes with our y scale because we take the log10 of negative values. (and taking the abs(head) results in making positive  a y axis while working with a plot on the negative y axis)

File: examples/07_diagnostic_plot.py [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+logarithmic derivative of the drawdown in a log-log plot.
+Often, this plot is used to identify the right approach for the aquifer estimations.
+
+"""
+
+
+import welltestpy as wtp
+
+campaign = wtp.load_campaign("Cmp_UFZ-campaign.cmp")
+campaign.diagnostic_plot("well_0", "well_1")
```
Comment (MuellerSeb) [inline examples/07_diagnostic_plot.py:15]: Ahh. You need to add an argument `rate` to `diagnostic_plot_pump_test` which is the pumping rate. When `rate<0` you have to multiply head by `-1` to get the "drawdown" and not the "piezometric level". Then, in the method `PumpingTest.diagnostic_plot` you need to pass `self.pumpingrate`.

**Review by MuellerSeb — CHANGES_REQUESTED** [review]
Almost there.

**Review by MuellerSeb — CHANGES_REQUESTED** [review]
Almost there. Please run `black -l 79 .` again.

**Comment by JarnoHerr** [comment]
I did before pushing. I tried again but it stays unchanged

**Comment by MuellerSeb** [comment]
> I did before pushing. I tried again but it stays unchanged

Ah, I guess it is because I added changes to `develop` in the meantime, where the configuration for `black` is now stored in `pyproject.toml`, which not yet present in your fork. But the CI workflow is taken from the current develop. So black is running there without the `-l 79` part, which results in the error. So let's ignore this error for now.

So I guess we are done here. When you say it is ready to merge, I am happy to accept this pull request :wink:

BTW: here you can have a look at the generated documentation from this pull request:
https://geostat-framework--14.org.readthedocs.build/projects/welltestpy/en/14/

**Review by MuellerSeb — APPROVED** [review]
Nice! That was really great work and a fruitful collaboration! Guess we both learned something :-)

**Comment by MuellerSeb** [comment]
We now have a merge conflict, that I will solve later. Thanks for the contribution, will merge ASAP.

**Review by MuellerSeb — APPROVED** [review]
LGTM
